### PR TITLE
Ensure versioning transition properly reschedules tasks with statebased replication

### PR DESCRIPTION
**Pull request by ShahabT**
## What changed?
Ensuring the activity and workflow task re-generation works fine on the passive side when statebased replication is turned on.

## Why?
Otherwise, the workflows can get stuck after failover in certain corner case.

## How did you test it?
Existing tests should cover regressions.
Will add new tests to ensure the fix works as intended.

- [ ] built
- [ ] run locally and tested manually
- [ ] covered by existing tests
- [ ] added new unit test(s)
- [ ] added new functional test(s)

## Potential risks
Patch candidate.

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -3143,7 +3143,7 @@ func (ms *MutableStateImpl) ApplyBuildIdRedirect(
 	ms.GetExecutionInfo().BuildIdRedirectCounter = redirectCounter
 
 	// Re-scheduling pending workflow and activity tasks.
-	err = ms.reschedulePendingWorkflowTask(false)
+	err = ms.reschedulePendingWorkflowTask()
```
Comment (ShahabT) [inline service/history/workflow/mutable_state_impl.go:3146]: these should be true but was scoped out from @stephanos PR based on https://github.com/temporalio/temporal/pull/8103/files#r2403627911

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -8559,14 +8578,12 @@ func (ms *MutableStateImpl) reschedulePendingWorkflowTask(invalidatePendingTasks
 		return nil
 	}
 
-	if invalidatePendingTasks {
-		// Increase the stamp ("version") to invalidate the pending non-speculative WFT.
-		// We don't invalidate speculative WFTs because they are very latency sensitive.
-		ms.executionInfo.WorkflowTaskStamp += 1
+	// Increase the stamp ("version") to invalidate the pending non-speculative WFT.
+	// We don't invalidate speculative WFTs because they are very latency sensitive.
+	ms.executionInfo.WorkflowTaskStamp += 1
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:8585]: let's set ms.workflowTaskUpdated = true since we are not using UpdateWorkflowTask() method here.

File: service/history/workflow/mutable_state_impl.go
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:3151]: nit for later: should this also somehow reuse reschedulePendingActivities as well?

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go (hunk 1 above)
Comment (stephanos) [reply service/history/workflow/mutable_state_impl.go:3146]: Right; I didn't want to make that change since I was worried it would require more reviews and introduce more risk.

File: service/history/workflow/mutable_state_impl.go
Comment (ShahabT) [reply service/history/workflow/mutable_state_impl.go:3151]: old versioning stuff will be cleaned up
